vllm/core/block/cpu_offloading_block_allocator.py

```diff
@@ -0,0 +1,400 @@
+             dst_device: Device) -> Dict[int, int]:
+
+        raise NotImplementedError("CPU offloading block allocator only "
+                                  "support preemption by recomputation.")
+
+    def _is_gpu_block(self, block_id: int) -> bool:
+        return block_id in self._allocators[Device.GPU].all_block_ids
+
+    def _is_gpu_block_unsafe(self, block_id: int) -> bool:
+        """Faster version of `_is_gpu_block` that doesn't check the block ID.
+        But assumes the that the block IDs are assigned contiguously, with GPU 
+        block IDs coming before the CPU block IDs.
+        """
+        return block_id < self.num_gpu_blocks
+
+    def _get_physical_block_id_unsafe(self, block_id: int) -> int:
+        """Returns the physical block ID of the given block ID.
+
+        This function avoids using the `allocator.get_physical_block_id()`
+        which is slow (O(NlogN)). Instead, this is based on the assumption
+        that the block IDs are assigned contiguously, with GPU block IDs coming
+        before CPU block IDs.
+
+        Args:
+            block_id (int): The block ID to get the physical block ID of.
+
+        Returns:
+            int: The physical block ID of the given block ID.
+
+        Note:
+            Please see the implementation of 
+            `CpuOffloadingBlockAllocator.create` for how the block IDs are
+            assigned.
+        """
+        if self._is_gpu_block_unsafe(block_id):
+            return block_id
+        else:
+            return block_id - self.num_gpu_blocks
+
+    def get_and_reset_swaps(self,
+                            now: float) -> Tuple[List[Tuple[int, int]], ...]:
+        """Returns and clears the mapping of source to destination block IDs.
+        Will be called right before scheduler step finishes.
+        
+        This function will do the following things:
+            1. Iterate over uncached blocks and see if we can copy it to CPU
+            2. Update all allocated CPU block time stamp
+            3. Free CPU blocks
+            4. Return and clear all swapping status
+            
+        Args:
+            now (float): The time stamp used to update CPU access time, so 
+            that CPU evictor can work.
+        
+        Returns:
+            A tuple of two lists: (blocks_to_swap_out, blocks_to_swap_in).
+            Each list is a List[Tuple[int, int]], containing the mapping of 
+            source to destination block IDs. The block IDs are physical block
+            IDs and it's expected to be used by the cache engine directly.
+        """
+
+        allocator = self._allocators[Device.GPU]
+        cpu_allocator = self._allocators[Device.CPU]
+
+        new_uncached_blocks: Deque[Block] = deque()
+
+        while self._uncached_blocks:
+            block = self._uncached_blocks.pop()
+            block_id = block.block_id
+
+            # check if this block is freed
+            if block_id is None:
```

Comment:
Because the block structure in _uncached_blocks is from a block memory pool, the block can be freed and allocated as a new block_id. However, while the block_id is allocated as a new block_id, this may cause some mistakes (or hurt performance). So, maybe we can remove the block in _uncached_blocks while block_manager calls the free function in cpu_offloading_block_allocator?

Reply:
I see what you mean. This makes sense. Let me think what's the best way to fix it. Iterating over the `_uncache_blocks` during free() will be too expensive IMO.

Reply:
@lixiaobai09 Hey, I've fixed this. Please see #11532